lib/Clam/CfgBuilder.cc

```diff
@@ -320,10 +322,90 @@ static Optional<lin_cst_t> cmpInstToCrabInt(CmpInst &I, crabLitFactory &lfac,
     if (I.getPredicate() == CmpInst::ICMP_ULT) {
       cst.set_unsigned();
     }
-    return cst;
+    return {cst, llvm::None};
     break;
   }
   case CmpInst::ICMP_ULE:
+  case CmpInst::ICMP_ULT: {
+    // translating unsigned comparisons by using select operators
+    // x cmp y, where cmp will be either <= or <
+    // will be translated into:
+    // if x >= 0
+    //     if y >= 0 then x cmp y
+    //     else // x >= 0 y < 0
+    //        false
+    // else
+    //     if y >= 0 then // x < 0 y >= 0
+    //        true
+    //     else x cmp y
+    var_t res = lfac.mkBoolVar();
+    lin_cst_t cmp;
+    if (I.getPredicate() == CmpInst::ICMP_ULE)
+      cmp = lin_cst_t(op0 <= op1);
+    else
+      cmp = lin_cst_t(op0 <= op1 - number_t(1));
+    lin_cst_t cst_op0 = lin_cst_t(op0 >= 0);
+    lin_cst_t cst_op1 = lin_cst_t(op1 >= 0);
+    if (cst_op0.is_tautology()) {
+      if (cst_op1.is_tautology())
+        bb.bool_assign(res, cmp);
+      else if (cst_op1.is_contradiction())
+        bb.bool_assign(res, lin_cst_t::get_false());
+      else {
+        var_t select_cond_1 = lfac.mkBoolVar();
+        bb.bool_assign(select_cond_1, cst_op1);
+        var_t cmp_var = lfac.mkBoolVar();
+        bb.bool_assign(cmp_var, cmp);
+        var_t v_false = lfac.mkBoolVar();
+        bb.bool_assign(v_false, lin_cst_t::get_false());
+        bb.bool_select(res, select_cond_1, cmp_var, v_false);
+      }
+    }
+    else if (cst_op0.is_contradiction()) {
+      if (cst_op1.is_tautology())
+        bb.bool_assign(res, lin_cst_t::get_true());
+      else if (cst_op1.is_contradiction())
+        bb.bool_assign(res, cmp);
+      else {
+        var_t select_cond_1 = lfac.mkBoolVar();
+        bb.bool_assign(select_cond_1, cst_op1);
+        var_t v_true = lfac.mkBoolVar();
+        bb.bool_assign(v_true, lin_cst_t::get_true());
+        var_t cmp_var = lfac.mkBoolVar();
+        bb.bool_assign(cmp_var, cmp);
+        bb.bool_select(res, select_cond_1, v_true, cmp_var);
+      }
+    }
+    else {
```

Comment:
Add a comment that says that this is the general case where we don't know at translation time whether operands are positive or negative